internal/app/readycheck.go

```diff
@@ -1,13 +1,161 @@
+	etcdConnectionTimeout = 5 * time.Second
+	etcdGetTimeout        = 5 * time.Second
+	etcdQueryInterval     = 2 * time.Second
+	etcdEndpointAddress   = ":2379"
+	schemeHTTP            = "http"
+	schemeHTTPS           = "https"
+)
 
-// SetupReadinessProbe sets up the readiness probe for this application.
+var (
+	etcdReady bool
+)
+
+// Create a struct which will hold the last status for etcd.
+// In SetupReadinessProbe first call go a.queryAndUpdateEtcdReadiness()/queryAndUpdateEtcdReadiness this function will
+// periodically query etcd and updates the readiness struct. The handler just reads from the struct.
+// If the struct has no status it assumes NotAvailable else it returns the current status.
+
+// SetupReadinessProbe sets up the readiness probe for this application. It is a blocking function and therefore
+// the consumer should always call this within a go-routine unless the caller itself wants to block on this which is unlikely.
 func (a *Application) SetupReadinessProbe() {
-	//the caller should call this in a go-routine as this method will setup a HTTP router and call ListenAndServe which blocks.
-	//If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	// Start go routine to regularly query etcd to update its readiness status.
+	go a.queryAndUpdateEtcdReadiness()
+	// If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	if a.isTLSEnabled() {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServeTLS(fmt.Sprintf(":%d", ReadyServerPort), a.cfg.ClientTLSInfo.CertFile, a.cfg.ClientTLSInfo.KeyFile, nil)
+		if err != nil {
+			a.logger.Fatal("failed to start TLS readiness endpoint", zap.Error(err))
+		}
+	} else {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServe(fmt.Sprintf(":%d", ReadyServerPort), nil)
+		if err != nil {
+			a.logger.Fatal("failed to start readiness endpoint", zap.Error(err))
+		}
+	}
+}
+
+// queryAndUpdateEtcdReadiness queries the etcd DB and updates the status of the query into the etcdStatus struct
+func (a *Application) queryAndUpdateEtcdReadiness() {
+	for {
+		etcdReady = a.isEtcdReady()
+		select {
+		case <-a.ctx.Done():
+			a.logger.Error("stopped periodic DB query: context cancelled", zap.Error(a.ctx.Err()))
+			return
+		case <-time.After(etcdQueryInterval):
+		}
+	}
+}
+
+// isEtcdReady checks if ETCD is ready by making a `GET` call (with a timeout).
+// if there is an error then it returns false else it returns true.
+func (a *Application) isEtcdReady() bool {
+	etcdConnCtx, cancelFunc := context.WithTimeout(a.ctx, etcdGetTimeout)
+	defer cancelFunc()
+	_, err := a.etcdClient.Get(etcdConnCtx, "foo", clientv3.WithSerializable())
+	if err != nil {
+		a.logger.Error("failed to retrieve from etcd db", zap.Error(err))
+	}
+	return err == nil
```

Comment:
Both are doing the same thing.

Reply:
I proposed this suggestion, because of it is more declarative way of saying returning `false` if any error or returning `true` than smart way `return err == nil`.

Just cosmetic change. Pls ignore it if you feel current approach better.